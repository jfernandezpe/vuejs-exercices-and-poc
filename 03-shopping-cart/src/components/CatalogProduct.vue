<template>
  <article class="card catalog-product">
    <h3>{{ name }}</h3>
    <img :src="image" width="300" />
    <p>{{ price }}</p>
    <button class="button" @click="addToCartButtonClicked" :disabled="disabled">
      Add to cart
    </button>
  </article>
</template>

<script>
export default {
  name: "CatalogProduct",
  props: {
    name: String,
    image: String,
    price: Number,
    disabled: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    addToCartButtonClicked() {
      if (!this.disabled) {
        this.$emit("addToCart");
      }
    }
  }
};
</script>

<style scoped lang="scss">
article {
  display: fex;
  flex-direction: column;
}
</style>
