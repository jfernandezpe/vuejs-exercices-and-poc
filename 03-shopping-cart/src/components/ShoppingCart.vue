<template>
  <article>
    <ul>
      <li v-for="product in cartProducts" :key="product.id">
        {{ product.name }} {{ product.price }}
        <button class="removeFromCart" @click="removeFromCart(product.id)">
          Remove
        </button>
      </li>
    </ul>
    <p>Total price: {{ totalPrice }}</p>
    <button class="buy" @click="buy">Buy</button>
  </article>
</template>

<script>
export default {
  name: "ShoppingCart",
  props: {
    cartProducts: {
      type: Array,
      default: () => []
    }
  },
  computed: {
    totalPrice() {
      return this.cartProducts.reduce((total, { price }) => {
        return total + price;
      }, 0);
    }
  },
  methods: {
    removeFromCart(idToRemove) {
      this.$emit("removeFromCart", idToRemove);
    },
    buy() {
      this.$emit("buy");
    }
  }
};
</script>

<style lang="sass" scoped></style>
