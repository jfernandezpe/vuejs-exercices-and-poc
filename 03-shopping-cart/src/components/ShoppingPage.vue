<template>
  <div class="shopping-page">
    <ShoppingCart
      class="shopping-cart"
      :cartProducts="cartProducts"
      v-on:removeFromCart="removeFromCart"
      v-on:buy="buy"
    />
    <ul>
      <li v-for="product in parsedProduct" :key="product.id">
        <CatalogProduct
          class="catalog-product"
          :name="product.name"
          :image="product.image"
          :price="product.price"
          :disabled="product.disabled"
          v-on:addToCart="addToCart(product.id)"
        />
      </li>
    </ul>
  </div>
</template>

<script>
import CatalogProduct from "./CatalogProduct.vue";
import ShoppingCart from "./ShoppingCart.vue";

export default {
  name: "ShoppingPage",
  components: { CatalogProduct, ShoppingCart },
  computed: {
    parsedProduct() {
      return this.products.map(({ id, name, image, price, disabled }) => {
        return {
          id,
          name,
          image,
          price: parseFloat(price),
          disabled: !!disabled
        };
      });
    }
  },
  methods: {
    addToCart(idToAdd) {
      //TODO: refactor and extract function
      this.products = this.products.map(product => {
        if (product.id === idToAdd) {
          return {
            ...product,
            disabled: true
          };
        }
        return product;
      });
      this.cartProducts = [
        ...this.cartProducts,
        this.parsedProduct.find(({ id }) => id === idToAdd)
      ];
    },
    removeFromCart(idToRemove) {
      this.products = this.products.map(product => {
        if (product.id === idToRemove) {
          return { ...product, disabled: false };
        }
        return product;
      });
      this.cartProducts = this.cartProducts.filter(
        ({ id }) => id !== idToRemove
      );
    },
    buy() {
      this.products = this.products.map(product => {
        return {
          ...product,
          disabled: false
        };
      });
      this.cartProducts = [];
    }
  },
  data() {
    return {
      cartProducts: [],
      products: [
        {
          id: "01",
          name: "Guitarra eléctrica",
          image: "https://picsum.photos/id/145/640/480",
          price: "1200.00"
        },
        {
          id: "02",
          name: "iPhone",
          image: "https://picsum.photos/id/160/640/480",
          price: "1800.00"
        },
        {
          id: "03",
          name: "Reloj antiguo",
          image: "https://picsum.photos/id/175/640/480",
          price: "200.00"
        },
        {
          id: "04",
          name: "Cámara Vintage",
          image: "https://picsum.photos/id/250/640/480",
          price: "950.00"
        },
        {
          id: "05",
          name: "Taza Cubana",
          image: "https://picsum.photos/id/30/640/480",
          price: "99.00"
        },
        {
          id: "06",
          name: "Amazón Kindle",
          image: "https://picsum.photos/id/367/640/480",
          price: "99.00"
        }
      ]
    };
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
.shopping-page {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  .shopping-cart {
    min-width: 25%;
  }
}

ul {
  display: flex;
  flex-wrap: wrap;
}
li {
  list-style-type: none;
}
</style>
