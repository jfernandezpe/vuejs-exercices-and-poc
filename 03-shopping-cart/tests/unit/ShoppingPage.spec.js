import { shallowMount } from "@vue/test-utils";
import ShoppingPage from "@/components/ShoppingPage.vue";
import CatalogProduct from "@/components/CatalogProduct.vue";
import ShoppingCart from "@/components/ShoppingCart.vue";

describe("ShoppingPage.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(ShoppingPage, {}); //TODO: shallow mount or mount?
  });
  it("shoult render the layout", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  describe("when receive addToCart event", () => {
    let child;
    beforeEach(async () => {
      child = wrapper.findAllComponents(CatalogProduct).wrappers[2];
    });
    describe("when the cart was empty", () => {
      beforeEach(async () => {
        child.vm.$emit("addToCart");
        await wrapper.vm.$nextTick();
      });
      it("should add the item to the cart products", () => {
        expect(wrapper.vm.cartProducts).toEqual([
          {
            id: "03",
            name: "Reloj antiguo",
            image: "https://picsum.photos/id/175/640/480",
            price: 200.0,
            disabled: true
          }
        ]);
      });
      it("the item should be disabled", () => {
        expect(
          wrapper.vm.parsedProduct.find(product => product.id === "03").disabled
        ).toBeTruthy();
      });
      it("should render the list", () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe("when the cart has items", () => {
      let previousProduct;
      beforeEach(async () => {
        previousProduct = {
          id: "02",
          name: "iPhone",
          image: "https://picsum.photos/id/160/640/480",
          price: 1800.0
        };
        wrapper.setData({ cartProducts: [previousProduct] });

        child.vm.$emit("addToCart");
        await wrapper.vm.$nextTick();
      });
      it("should add the item to the previous products", () => {
        expect(wrapper.vm.cartProducts).toEqual([
          previousProduct,
          {
            id: "03",
            name: "Reloj antiguo",
            image: "https://picsum.photos/id/175/640/480",
            price: 200.0,
            disabled: true
          }
        ]);
      });
      it("should render the list", () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
  describe("when receive a remove from cart event", () => {
    const iPhoneId = "02";
    beforeEach(async () => {
      const products = [
        {
          id: "01",
          name: "Guitarra elÃ©ctrica",
          image: "https://picsum.photos/id/145/640/480",
          price: "1200.00",
          disabled: false
        },
        {
          id: iPhoneId,
          name: "iPhone",
          image: "https://picsum.photos/id/160/640/480",
          price: "1800.00",
          disabled: true
        }
      ];
      wrapper.setData({ products });
      await wrapper.vm.$nextTick();

      const [, parsedIPhone] = wrapper.vm.parsedProduct;

      wrapper.setData({ cartProducts: [parsedIPhone] });
      await wrapper.vm.$nextTick();

      //Assert
      const shoppingCart = wrapper.findComponent(ShoppingCart);
      shoppingCart.vm.$emit("removeFromCart", iPhoneId);
      await wrapper.vm.$nextTick();
    });
    it("should remove the product form the cart list", () => {
      expect(wrapper.vm.cartProducts).toStrictEqual([]);
    });
    it("should enable the product", () => {
      const iPhone = wrapper.vm.products.find(
        product => product.id === iPhoneId
      );
      expect(iPhone.disabled).toBeFalsy();
    });
  });

  describe("when receive a buy event", () => {
    beforeEach(async () => {
      const products = wrapper.vm.products.map(product => {
        return { ...product, disabled: true };
      });
      wrapper.setData({ products });
      wrapper.setData({ cartProducts: wrapper.vm.parsedProduct });
      const shoppingCart = wrapper.findComponent(ShoppingCart);
      shoppingCart.vm.$emit("buy");
      await wrapper.vm.$nextTick();
    });
    it("should reset the products", () => {
      wrapper.vm.parsedProduct.forEach(product => {
        expect(product.disabled).toBeFalsy();
      });
    });
    it("should reset the cartProducts", () => {
      expect(wrapper.vm.cartProducts).toEqual([]);
    });
    it("should render the layout", () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
