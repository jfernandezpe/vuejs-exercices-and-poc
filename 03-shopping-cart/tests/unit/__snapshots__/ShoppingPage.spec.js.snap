// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingPage.vue shoult render the layout 1`] = `
<div
  class="shopping-page"
>
  <shoppingcart-stub
    cartproducts=""
    class="shopping-cart"
  />
   
  <ul>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/145/640/480"
        name="Guitarra eléctrica"
        price="1200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/160/640/480"
        name="iPhone"
        price="1800"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/175/640/480"
        name="Reloj antiguo"
        price="200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/250/640/480"
        name="Cámara Vintage"
        price="950"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/30/640/480"
        name="Taza Cubana"
        price="99"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/367/640/480"
        name="Amazón Kindle"
        price="99"
      />
    </li>
  </ul>
</div>
`;

exports[`ShoppingPage.vue when receive a buy event should render the layout 1`] = `
<div
  class="shopping-page"
>
  <shoppingcart-stub
    cartproducts=""
    class="shopping-cart"
  />
   
  <ul>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/145/640/480"
        name="Guitarra eléctrica"
        price="1200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/160/640/480"
        name="iPhone"
        price="1800"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/175/640/480"
        name="Reloj antiguo"
        price="200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/250/640/480"
        name="Cámara Vintage"
        price="950"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/30/640/480"
        name="Taza Cubana"
        price="99"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/367/640/480"
        name="Amazón Kindle"
        price="99"
      />
    </li>
  </ul>
</div>
`;

exports[`ShoppingPage.vue when receive addToCart event when the cart has items should render the list 1`] = `
<div
  class="shopping-page"
>
  <shoppingcart-stub
    cartproducts="[object Object],[object Object]"
    class="shopping-cart"
  />
   
  <ul>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/145/640/480"
        name="Guitarra eléctrica"
        price="1200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/160/640/480"
        name="iPhone"
        price="1800"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        disabled="true"
        image="https://picsum.photos/id/175/640/480"
        name="Reloj antiguo"
        price="200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/250/640/480"
        name="Cámara Vintage"
        price="950"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/30/640/480"
        name="Taza Cubana"
        price="99"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/367/640/480"
        name="Amazón Kindle"
        price="99"
      />
    </li>
  </ul>
</div>
`;

exports[`ShoppingPage.vue when receive addToCart event when the cart was empty should render the list 1`] = `
<div
  class="shopping-page"
>
  <shoppingcart-stub
    cartproducts="[object Object]"
    class="shopping-cart"
  />
   
  <ul>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/145/640/480"
        name="Guitarra eléctrica"
        price="1200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/160/640/480"
        name="iPhone"
        price="1800"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        disabled="true"
        image="https://picsum.photos/id/175/640/480"
        name="Reloj antiguo"
        price="200"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/250/640/480"
        name="Cámara Vintage"
        price="950"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/30/640/480"
        name="Taza Cubana"
        price="99"
      />
    </li>
    <li>
      <catalogproduct-stub
        class="catalog-product"
        image="https://picsum.photos/id/367/640/480"
        name="Amazón Kindle"
        price="99"
      />
    </li>
  </ul>
</div>
`;
